File: packages/engine/src/scaffoldGolden.ts
import type { BindingInventory, BindingItem, Block, RunDocument } from "@runoff/core";
import { boundnessLine } from "./goldenBinding.js";

export interface ScaffoldGoldenInput {
  id: string;
  label: string;
  period: string | null;
  document: RunDocument;
  inventory: BindingInventory;
}
export interface ScaffoldQuery { name: string; sql: string; provenance: "verified" | "verified-mismatch" }
export interface ScaffoldSection { key: string; heading: string; prose: string; queries: ScaffoldQuery[]; warnings: string[] }
export interface ScaffoldDigest {
  goldenId: string;
  label: string;
  period: string | null;
  boundness: string;
  sections: ScaffoldSection[];
}

const PROSE_CAP = 1500;

/** Spec §2.3: spans joined with a space; tables header-only; blocks joined with \n; capped. */
function serializeProse(blocks: Block[]): string {
  const text = blocks
    .map((b) =>
      b.type === "paragraph"
        ? b.spans.map((sp) => sp.text).join(" ")
        : `[table ${b.columns.length} cols × ${b.rows.length} rows: ${b.columns.join(" | ")}]`,
    )
    .join("\n");
  return text.length > PROSE_CAP ? `${text.slice(0, PROSE_CAP)}…` : text;
}

/** Spec §2.2: one digest section per document section (document order); bound/mismatch SQL lifted, the rest becomes warnings. */
export function buildScaffoldDigest(g: ScaffoldGoldenInput): ScaffoldDigest {
  const bySection = new Map<string, BindingItem[]>();
  for (const it of g.inventory.items) {
    const list = bySection.get(it.anchor.sectionKey) ?? [];
    list.push(it);
    bySection.set(it.anchor.sectionKey, list);
  }
  const sections = g.document.sections.map((s) => {
    const queries: ScaffoldQuery[] = [];
    const warnings: string[] = [];
    const used = new Set<string>();
    for (const it of bySection.get(s.key) ?? []) {
      const b = it.binding;
      if (b && (b.status === "bound" || b.status === "mismatch")) {
        let name = it.id;
        for (let i = 2; used.has(name); i++) name = `${it.id}_${i}`;
        used.add(name);
        queries.push({ name, sql: b.sql, provenance: b.status === "bound" ? "verified" : "verified-mismatch" });
        if (b.status === "mismatch") {
          warnings.push(`"${it.raw}" mismatches current data (golden says ${it.raw} · data ${String(b.verifiedValue)})`);
        }
      } else {
        warnings.push(`"${it.raw}" has no data backing (${it.reason ?? "unbound"})`);
      }
    }
    if (queries.length === 0) warnings.push("no verified queries in this section");
    return { key: s.key, heading: s.heading, prose: serializeProse(s.blocks), queries, warnings };
  });
  return { goldenId: g.id, label: g.label, period: g.period, boundness: boundnessLine(g.inventory), sections };
}

/** Spec §2.4. SQL is NOT truncated here — the agent lifts it verbatim. */
export function renderScaffoldDigest(d: ScaffoldDigest): string {
  const lines = [`SCAFFOLD DIGEST — golden "${d.label}" (period ${d.period ?? "none"}, ${d.boundness})`];
  for (const s of d.sections) {
    lines.push("", `## section: ${s.key} — ${s.heading}`, "prose:", s.prose);
    if (s.queries.length) {
      lines.push("queries:");
      for (const q of s.queries) lines.push(`  ${q.name}: ${q.sql}  [${q.provenance}]`);
    }
    if (s.warnings.length) {
      lines.push("warnings:");
      for (const w of s.warnings) lines.push(`  - ${w}`);
    }
  }
  return lines.join("\n");
}

File: backend/runoff_api/engine/scaffold_golden.py
"""Port of packages/engine/src/scaffoldGolden.ts —
`serializeProse`, `buildScaffoldDigest`, `renderScaffoldDigest`.

Deterministic scaffold-digest engine (no LLM): groups an inventory's bound /
mismatch bindings into per-section queries and turns the rest into warnings,
then renders the digest text an agent lifts SQL from verbatim.

Everything flows as plain dicts with camelCase keys (the TS runtime shape).
"""

from runoff_api.services.golden_binding import _js_string, boundness_line

PROSE_CAP = 1500


def serialize_prose(blocks: list[dict]) -> str:
    """Spec §2.3: spans joined with a space; tables header-only; blocks joined with \\n; capped."""
    text = "\n".join(
        " ".join(sp["text"] for sp in b["spans"])
        if b["type"] == "paragraph"
        else f"[table {len(b['columns'])} cols × {len(b['rows'])} rows: {' | '.join(b['columns'])}]"
        for b in blocks
    )
    return f"{text[:PROSE_CAP]}…" if len(text) > PROSE_CAP else text


def build_scaffold_digest(g: dict) -> dict:
    """Spec §2.2: one digest section per document section (document order);
    bound/mismatch SQL lifted, the rest becomes warnings.
    """
    by_section: dict[str, list[dict]] = {}
    for it in g["inventory"]["items"]:
        by_section.setdefault(it["anchor"]["sectionKey"], []).append(it)
    sections = []
    for s in g["document"]["sections"]:
        queries: list[dict] = []
        warnings: list[str] = []
        used: set[str] = set()
        for it in by_section.get(s["key"], []):
            b = it["binding"]
            if b and b["status"] in ("bound", "mismatch"):
                name = it["id"]
                i = 2
                while name in used:
                    name = f"{it['id']}_{i}"
                    i += 1
                used.add(name)
                queries.append(
                    {
                        "name": name,
                        "sql": b["sql"],
                        "provenance": "verified" if b["status"] == "bound" else "verified-mismatch",
                    }
                )
                if b["status"] == "mismatch":
                    warnings.append(
                        f'"{it["raw"]}" mismatches current data '
                        f'(golden says {it["raw"]} · data {_js_string(b["verifiedValue"])})'
                    )
            else:
                reason = it["reason"] if it.get("reason") is not None else "unbound"
                warnings.append(f'"{it["raw"]}" has no data backing ({reason})')
        if len(queries) == 0:
            warnings.append("no verified queries in this section")
        sections.append(
            {
                "key": s["key"],
                "heading": s["heading"],
                "prose": serialize_prose(s["blocks"]),
                "queries": queries,
                "warnings": warnings,
            }
        )
    return {
        "goldenId": g["id"],
        "label": g["label"],
        "period": g["period"],
        "boundness": boundness_line(g["inventory"]),
        "sections": sections,
    }


def render_scaffold_digest(d: dict) -> str:
    """Spec §2.4. SQL is NOT truncated here — the agent lifts it verbatim."""
    period = d["period"] if d["period"] is not None else "none"
    lines = [f'SCAFFOLD DIGEST — golden "{d["label"]}" (period {period}, {d["boundness"]})']
    for s in d["sections"]:
        lines.extend(["", f'## section: {s["key"]} — {s["heading"]}', "prose:", s["prose"]])
        if s["queries"]:
            lines.append("queries:")
            for q in s["queries"]:
                lines.append(f'  {q["name"]}: {q["sql"]}  [{q["provenance"]}]')
        if s["warnings"]:
            lines.append("warnings:")
            for w in s["warnings"]:
                lines.append(f"  - {w}")
    return "\n".join(lines)
